syntax = "proto3";

package engine;

// Engine Service - Unified interface for all engine types (vLLM, etc.)
service EngineService {
  // Health check
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
  
  // Generate text (non-streaming)
  rpc Generate(GenerateRequest) returns (GenerateResponse);
  
  // Generate text (streaming)
  rpc GenerateStream(GenerateRequest) returns (stream GenerateStreamResponse);
  
  // Chat completion (non-streaming)
  rpc Chat(ChatRequest) returns (ChatResponse);
  
  // Chat completion (streaming)
  rpc ChatStream(ChatRequest) returns (stream ChatStreamResponse);
  
  // Create embeddings
  rpc CreateEmbedding(EmbeddingRequest) returns (EmbeddingResponse);
  
  // Cancel a request
  rpc CancelRequest(CancelRequest) returns (CancelResponse);
  
  // Get metrics
  rpc GetMetrics(MetricsRequest) returns (MetricsResponse);
  
  // Get KV cache status
  rpc GetKVCacheStatus(KVCacheStatusRequest) returns (KVCacheStatusResponse);
  
  // Shutdown engine
  rpc Shutdown(ShutdownRequest) returns (ShutdownResponse);
  
  // Get model info
  rpc GetModelInfo(ModelInfoRequest) returns (ModelInfoResponse);
}

// Health Check
message HealthCheckRequest {}

message HealthCheckResponse {
  string status = 1;
  string version = 2;
}

// Generate
message GenerateRequest {
  string prompt = 1;
  optional int32 max_tokens = 2;
  optional float temperature = 3;
  optional float top_p = 4;
  optional float top_k = 5;
  optional float presence_penalty = 6;
  optional float frequency_penalty = 7;
  repeated string stop = 8;
  optional int32 n = 9;
  optional bool stream = 10;
  optional string session_id = 11;
  optional int64 seed = 12;
}

message GenerateResponse {
  string id = 1;
  string text = 2;
  int32 tokens_generated = 3;
  string finish_reason = 4;
  Usage usage = 5;
}

message GenerateStreamResponse {
  string id = 1;
  string text = 2;
  bool is_finished = 3;
  optional string finish_reason = 4;
  optional Usage usage = 5;
}

// Chat
message ChatRequest {
  repeated ChatMessage messages = 1;
  optional int32 max_tokens = 2;
  optional float temperature = 3;
  optional float top_p = 4;
  optional float top_k = 5;
  optional float presence_penalty = 6;
  optional float frequency_penalty = 7;
  repeated string stop = 8;
  optional int32 n = 9;
  optional bool stream = 10;
  optional string session_id = 11;
  optional int64 seed = 12;
  optional string response_format = 13; // "json_object" for structured output
}

message ChatMessage {
  string role = 1;
  string content = 2;
  optional string name = 3;
}

message ChatResponse {
  string id = 1;
  string object = 2;
  int64 created = 3;
  string model = 4;
  repeated ChatChoice choices = 5;
  Usage usage = 6;
}

message ChatChoice {
  int32 index = 1;
  ChatMessage message = 2;
  string finish_reason = 3;
}

message ChatStreamResponse {
  string id = 1;
  string object = 2;
  int64 created = 3;
  string model = 4;
  repeated ChatStreamChoice choices = 5;
  optional Usage usage = 6;
}

message ChatStreamChoice {
  int32 index = 1;
  ChatMessageDelta delta = 2;
  optional string finish_reason = 3;
}

message ChatMessageDelta {
  optional string role = 1;
  optional string content = 2;
}

// Embedding
message EmbeddingRequest {
  oneof input {
    string text = 1;
    StringList texts = 2;
  }
  optional string encoding_format = 3;
}

message StringList {
  repeated string values = 1;
}

message EmbeddingResponse {
  string object = 1;
  repeated Embedding data = 2;
  string model = 3;
  Usage usage = 4;
}

message Embedding {
  string object = 1;
  repeated float embedding = 2;
  int32 index = 3;
}

// Usage
message Usage {
  int32 prompt_tokens = 1;
  int32 completion_tokens = 2;
  int32 total_tokens = 3;
}

// Cancel
message CancelRequest {
  string request_id = 1;
}

message CancelResponse {
  bool success = 1;
  string message = 2;
}

// Metrics
message MetricsRequest {}

message MetricsResponse {
  int32 num_requests_running = 1;
  int32 num_requests_waiting = 2;
  double gpu_cache_usage_perc = 3;
  double cpu_cache_usage_perc = 4;
  int64 num_total_tokens = 5;
  double throughput_tokens_per_sec = 6;
}

// KV Cache Status
message KVCacheStatusRequest {}

message KVCacheStatusResponse {
  int64 num_blocks_total = 1;
  int64 num_blocks_free = 2;
  int64 num_blocks_used = 3;
  double utilization_perc = 4;
}

// Shutdown
message ShutdownRequest {}

message ShutdownResponse {
  bool success = 1;
}

// Model Info
message ModelInfoRequest {}

message ModelInfoResponse {
  string model_name = 1;
  string model_path = 2;
  string dtype = 3;
  int32 max_model_len = 4;
  int32 num_layers = 5;
  int32 hidden_size = 6;
  int32 num_attention_heads = 7;
  int32 vocab_size = 8;
}
